src/lightgbm/src/main/scala/TrainUtils.scala

```diff
@@ -36,95 +36,172 @@ private object TrainUtils extends Serializable {
-    if (!inputRows.hasNext)
-      List[LightGBMBooster]().toIterator
-
-    val rows = inputRows.toArray
+  def generateDataset(rows: Array[Row], labelColumn: String, featuresColumn: String,
+                      weightColumn: Option[String],
+                      referenceDataset: Option[SWIGTYPE_p_void]): Option[SWIGTYPE_p_void] = {
     val numRows = rows.length
     val labels = rows.map(row => row.getDouble(row.fieldIndex(labelColumn)))
     val hrow = rows.head
     var datasetPtr: Option[SWIGTYPE_p_void] = None
-    try {
-      datasetPtr =
-        if (hrow.get(hrow.fieldIndex(featuresColumn)).isInstanceOf[DenseVector]) {
-          val rowsAsDoubleArray = rows.map(row => row.get(row.fieldIndex(featuresColumn)) match {
-            case dense: DenseVector => dense.toArray
-            case sparse: SparseVector => sparse.toDense.toArray
-          })
-          Some(LightGBMUtils.generateDenseDataset(numRows, rowsAsDoubleArray))
-        } else {
-          val rowsAsSparse = rows.map(row => row.get(row.fieldIndex(featuresColumn)) match {
-            case dense: DenseVector => dense.toSparse
-            case sparse: SparseVector => sparse
-          })
-          Some(LightGBMUtils.generateSparseDataset(rowsAsSparse))
-        }
+    datasetPtr =
+      if (hrow.get(hrow.fieldIndex(featuresColumn)).isInstanceOf[DenseVector]) {
+        val rowsAsDoubleArray = rows.map(row => row.get(row.fieldIndex(featuresColumn)) match {
+          case dense: DenseVector => dense.toArray
+          case sparse: SparseVector => sparse.toDense.toArray
+        })
+        Some(LightGBMUtils.generateDenseDataset(numRows, rowsAsDoubleArray, referenceDataset))
+      } else {
+        val rowsAsSparse = rows.map(row => row.get(row.fieldIndex(featuresColumn)) match {
+          case dense: DenseVector => dense.toSparse
+          case sparse: SparseVector => sparse
+        })
+        Some(LightGBMUtils.generateSparseDataset(rowsAsSparse, referenceDataset))
+      }
+
+    // Validate generated dataset has the correct number of rows and cols
+    validateDataset(datasetPtr.get)
+    addField(labels, "label", numRows, datasetPtr)
+    if (weightColumn.isDefined) {
+      val weights = rows.map(row => row.getDouble(row.fieldIndex(weightColumn.get)))
+      addField(weights, "weight", numRows, datasetPtr)
+    }
+    datasetPtr
+  }
+
+  def createBooster(trainParams: TrainParams, trainDatasetPtr: Option[SWIGTYPE_p_void],
+                    validDatasetPtr: Option[SWIGTYPE_p_void]): Option[SWIGTYPE_p_void] = {
+    // Create the booster
+    val boosterOutPtr = lightgbmlib.voidpp_handle()
+    val parameters = trainParams.toString()
+    LightGBMUtils.validate(lightgbmlib.LGBM_BoosterCreate(trainDatasetPtr.get, parameters, boosterOutPtr), "Booster")
+    val boosterPtr = Some(lightgbmlib.voidpp_value(boosterOutPtr))
+    if (trainParams.modelString != null && !trainParams.modelString.isEmpty) {
```

Comment:
will the model string ever be empty? If so we should try to remove that logic as early as possible and work with options

Reply:
done, changed to option of string